packages/ra-core/src/dataProvider/useCreate.ts

```diff
@@ -106,91 +217,287 @@ export const useCreate = <
+                    queryClient.invalidateQueries({ queryKey });
+                });
+            }
+
+            if (callTimeOnSettled.current) {
+                return callTimeOnSettled.current(
+                    data,
+                    error,
+                    variables,
+                    context
+                );
+            }
+            if (mutationOptions.onSettled) {
+                return mutationOptions.onSettled(
+                    data,
+                    error,
+                    variables,
+                    context
+                );
             }
         },
     });
 
-    const create = (
+    const create = async (
         callTimeResource: string | undefined = resource,
         callTimeParams: Partial<CreateParams<Partial<RecordType>>> = {},
         callTimeOptions: MutateOptions<
             ResultRecordType,
             MutationError,
             Partial<UseCreateMutateParams<RecordType>>,
             unknown
-        > & { returnPromise?: boolean } = {}
+        > & { mutationMode?: MutationMode; returnPromise?: boolean } = {}
     ) => {
         const {
-            returnPromise = options.returnPromise,
+            mutationMode,
+            returnPromise = mutationOptions.returnPromise,
+            onError,
+            onSettled,
+            onSuccess,
             ...otherCallTimeOptions
         } = callTimeOptions;
 
-        hasCallTimeOnError.current = !!otherCallTimeOptions.onError;
-        hasCallTimeOnSuccess.current = !!otherCallTimeOptions.onSuccess;
-        hasCallTimeOnSettled.current = !!otherCallTimeOptions.onSettled;
+        // Store the mutation with middlewares to avoid losing them if the calling component is unmounted
+        if (getMutateWithMiddlewares) {
+            mutateWithMiddlewares.current = getMutateWithMiddlewares(
+                dataProvider.create.bind(dataProvider)
+            );
+        } else {
+            mutateWithMiddlewares.current = dataProvider.create;
+        }
+
+        // We need to keep the onSuccess callback here and not in the useMutation for undoable mutations
+        hasCallTimeOnSuccess.current = !!onSuccess;
+        // We need to store the onError and onSettled callbacks here to be able to call them in the useMutation hook
+        // so that they are called even when the calling component is unmounted
+        callTimeOnError.current = onError;
+        callTimeOnSettled.current = onSettled;
+
+        // store the hook time params *at the moment of the call*
+        // because they may change afterwards, which would break the undoable mode
+        // as the previousData would be overwritten by the optimistic update
+        paramsRef.current = params;
 
-        if (returnPromise) {
-            return mutation.mutateAsync(
+        if (mutationMode) {
+            mode.current = mutationMode;
+        }
+
+        if (returnPromise && mode.current !== 'pessimistic') {
```

Comment:
Is it needed ? as the returnPromise is defined base on the mode https://github.com/marmelab/react-admin/blob/35bf17e551f3fcf1acb171870e7b711f6b427ed0/packages/ra-core/src/controller/create/useCreateController.ts#L147